pyproject.toml

```diff
@@ -32,6 +32,7 @@ dependencies = [
     "httpx",
     "aiofiles==22.1.0",
     "lowtime",
+    "anyio<4" # why it downloads 4.2 if I put ths under test?
```

Comment:
This was for testing(pytest) and I needed to have less than version 4 for compatibility with fastapi.testclient. I tried to put that on the test dependency, but for some reason if I put that into the test dependency, in installs anyio with 4.2.0 version where I need is 3.7

Reply:
Hmm I'm honestly not sure why that happens. But let's try to fix that issue directly instead of trying to introduce a new dependency.

Reply:
Perhaps pin with `anyio==3.7.1`?

Reply:
I think anyio is getting installed as a part of `fastapi`
```
equirement already satisfied: uvicorn>=0.12.0 in /opt/homebrew/lib/python3.11/site-packages (from uvicorn[standard]>=0.12.0; extra == "all"->fastapi[all]==0.87.0->zeus-ml==0.8.0) (0.27.0.post1)
Collecting anyio<5,>=3.4.0 (from starlette==0.21.0->fastapi==0.87.0->fastapi[all]==0.87.0->zeus-ml==0.8.0)
```
This will install anyio 4.2. However, this will cause this error`AttributeError: module 'anyio' has no attribute 'start_blocking_portal'` from `testclient`. 

The reason for this error is that based on anyio's documentation, this is not supported 

https://anyio.readthedocs.io/en/stable/versionhistory.html#:~:text=%E2%86%92%20TaskGroup.start_soon()-,BACKWARDS%20INCOMPATIBLE,-start_blocking_portal()%20must

I think we should specify `anyio<4` in the main dependency. Not sure there is a better option.

Reply:
You mean FastAPI requires 4.2 but fastapi.testserver requires 3.7? I cannot imagine such discrepency in a framework like FastAPI. Look through their repo issues.

Reply:
anyio is only needed for testing, so it's not going to be part of the main dependency. This is the invariant. Please find a fix or a workaround.

Reply:
https://github.com/encode/starlette/issues/2262 
They recommend pinning or upgrading Fastapi to recent one.

Reply:
Yeah let’s upgrade

Reply:
Any reason we want to do `fastapi[all]`? We have a dependency conflict with pydantic when updating fastapi.
For `"fastapi[all]==0.103.1"` (lowest version that fixes anyio issue), we get 
```
zeus-ml 0.8.0 depends on pydantic<2
    fastapi 0.103.1 depends on pydantic!=1.8, !=1.8.1, !=2.0.0, !=2.0.1, !=2.1.0, <3.0.0 and >=1.7.4
    fastapi[all] 0.103.1 depends on pydantic!=1.8, !=1.8.1, !=2.0.0, !=2.0.1, !=2.1.0, <3.0.0 and >=1.7.4
    pydantic-extra-types 2.1.0 depends on pydantic>=2.0.3
```
pydantic-extra-types  is coming from fastapi 0.100.0 
https://fastapi.tiangolo.com/release-notes/#01032:~:text=When%20you%20install,now%20also%20includes

I remember updating pydantic would require us to fix quiet lots of code. What do you think?

Reply:
Oh shoot now I remembered why I pinned FastAPI. It's because of Pydantic. Can you find a reasonably old version of FastAPI that does not pin to pydantic 2.0?

Reply:
Pydantic 2.0 is still early, and not upgrading is because of our users. Many people just stay at Pydantic 1.0, and if we require 2.0 in any way, that will become a roadblock for them.

Reply:
If there's not an intermediate FastAPI version that fixes anyio and does not require pydantic 2.0, we want to consider isolating fastapi (and pydantic 2.0) in an optional dependency, like `bso`.

Reply:
Ah, fixed the issue. Since the `anyio` is the problem here, I moved `anyio < 4` to the test dependency.
Then, I tried `pip install -e ".[test]"` and it worked out smoothly. The reason I wasn't able to get it working before was I was using `pip install -e .[test]` 

Now all other main dependencies can stay the same as before.